### PR TITLE
Print all users if no arguments are passed

### PR DESCRIPTION
**Pull request by joshmoore**
Useful for debugging purposes.

### Before

```
(base) /opt/omero-ldaptool $build/install/omero-ldaptool/bin/omero-ldaptool cfg
Missing required parameter: <username>
Usage: <main class> [--debug] [--help] <config> <username>
      <config>     LDAP configuration properties file
      <username>   Username to search for
      --debug      Set logging level to DEBUG
      --help       Display this help and exit
```

### After

```
(base) /opt/omero-ldaptool $build/install/omero-ldaptool/bin/omero-ldaptool
Missing required parameters: <config>, <username>
Usage: <main class> [--debug] [--help] <config> <username>
      <config>     LDAP configuration properties file
      <username>   Username to search for (defaults to all users)
      --debug      Set logging level to DEBUG
      --help       Display this help and exit
(base) /opt/omero-ldaptool $build/install/omero-ldaptool/bin/omero-ldaptool cfg
2021-03-29 08:59:59,909 [main] WARN  ome.services.util.ReadOnlyStatus - not configured for read-only binary repository but assuming read-only anyway
---
- u1:  dn: uid=u1,ou=Users,dc=openmicroscopy,dc=org
  groups:
- u2:  dn: uid=u2,ou=Users,dc=openmicroscopy,dc=org
  groups:
```

### CODE REVIEW

**Comment by chris-allan** [comment]
Not necessarily adverse to the "println" style but that's not what we have been doing so far here printing to INFO. Part of the design, in particular `LoggingLdapTemplate`, is specifically set up to log LDAP queries and this changes a lot of those defaults.

/cc @stick

**Comment by joshmoore** [comment]
Happy to take guidance. If the logging should stay at INFO, then I'd likely introduce an option to get back this parseable stdout.

**Comment by chris-allan** [comment]
@stick is our biggest user of the tool so will definitely get his feedback and go from there. Regardless of which way we go it should definitely be consistent all users vs. one.

Thoughts with respect to CSV / JSON as output to aid usability?

**Comment by stick** [comment]
To be honest returning all users in the ldap context can be dangerous, some of the orgs we use this with have 10,000+ users/uids in ldap.

I/we don't use this tool to gather data so output format is mostly inconsequential (to me), it's use is almost entirely diagnostic or exploratory.  Making sure credentials are correct, making sure mapping fields are accurate, making sure groups are mapping as desired etc, do we need global catalog, etc.

That aside CSV / JSON would be fine.

**Comment by joshmoore** [comment]
Updated to print out group info. https://stackoverflow.com/questions/36449067/how-to-paginate-in-spring-ldap-using-skip-value might be a path to at least not print _every_ user, though I assume paging through the users is out of scope.

**Comment by chris-allan** [comment]
👍 

Maybe from a safety perspective in the short term we continue with `<username>` but do nothing unless you pass `--all` or similar? This would ensure people don't get themselves banned from accessing their LDAP directories by making huge queries.

With respect to the output, can we just make it the same regardless of whether we're in single or multiple user mode? If people want to get more details during debugging `--debug` then does more or less what it does now and we don't have this disconnect between `INFO` and `WARN` levels.

**Comment by joshmoore** [comment]
Done:
```
(base) /opt/omero-ldaptool $build/install/omero-ldaptool/bin/omero-ldaptool cfg u1
---
- dn: uid=u1,ou=Users,dc=openmicroscopy,dc=org
  omeName: u1
  id: null
  firstName: J.
  middleName: null
  lastName: Doe
  email: null
  institution: null
  ldap: true
  owner: []
  member: [1]
(base) /opt/omero-ldaptool $build/install/omero-ldaptool/bin/omero-ldaptool cfg u1 --all
---
- dn: uid=u1,ou=Users,dc=openmicroscopy,dc=org
  omeName: u1
  id: null
  firstName: J.
  middleName: null
  lastName: Doe
  email: null
  institution: null
  ldap: true
  owner: []
  member: [1]
(base) /opt/omero-ldaptool $build/install/omero-ldaptool/bin/omero-ldaptool cfg
---
usage: CONFIGFILE USERNAME
       CONFIGFILE --all
```

**Comment by chris-allan** [comment]
Looks great. We'll test with @stick on a couple deployments and then get this merged.

**Comment by chris-allan** [comment]
With picocli I wonder if we can make a positional argument and another argument mutually exclusive. Did you look into that at all @joshmoore? Might have to update our picocli version.

**Comment by chris-allan** [comment]
* https://picocli.info/#_mutually_exclusive_options
 
 Might be an idea to do `--all | --user`? Would save you having to deal with printing and handling usage specifically for those cases?

**Comment by joshmoore** [comment]
No objections from my side if we want to change the previous behavior. I was initially trying to be conservative, but that's largely flown out the window at this point ;)

**Comment by chris-allan** [comment]
Agreed. Yeah, give it a look. I'll chat with @stick and see if he has any big concerns.

**Comment by chris-allan** [comment]
Since the PR changes the command line interface quite substantially I'll release this as 0.2.0 when we're ready.

/cc @stick

**Comment by joshmoore** [comment]
Forgot to check this one, but definitely looks nicer now (with shell colors):

```
(base) /opt/omero-ldaptool $build/install/omero-ldaptool/bin/omero-ldaptool cfg
Error: Missing required argument (specify one of these): (--all | --user=<username>)
Usage: <main class> [--debug] [--help] (--all | --user=<username>) <config>
      <config>            LDAP configuration properties file
      --all               Print all users
      --debug             Set logging level to DEBUG
      --help              Display this help and exit
      --user=<username>   Username to search
```
:+1:

**Comment by chris-allan** [comment]
Awesome. Looks great.

File: src/main/java/com/glencoesoftware/ldaptool/Main.java [hunk 1]
```diff
@@ -112,58 +123,81 @@ public GroupLoader newGroupLoader(
...
-        List<Long> ownedGroupIds = groupLoader.getOwnedGroups();
-        log.info(
-            "Would be member of Group IDs={}",
-            Arrays.toString(groupIds.toArray())
-        );
-        log.info(
-            "Would be owner of Group IDs={}",
-            Arrays.toString(ownedGroupIds.toArray())
-        );
+        System.out.println("- dn: " + dn);
```
Comment (joshmoore) [inline src/main/java/com/glencoesoftware/ldaptool/Main.java]: These really should be in quotes. Any objections?

**Comment by stick** [comment]
Only printing an array index for groups is problematic (based on the output example), I know we don't have real group ids from the db but what's useful is the ldap group names that end up matching.

We're not using this anywhere programmatically etc, and only 1 case I know of where someone other than myself is using it so change in usage is fine no issues from me.

File: src/main/java/com/glencoesoftware/ldaptool/Main.java (hunk 1 above)
Comment (chris-allan) [reply src/main/java/com/glencoesoftware/ldaptool/Main.java]: None. Completely up to you.

**Comment by chris-allan** [comment]
@stick: Would love to do that. Unfortunately that's not possible with the way the code is currently structured. IDs are what we have.

**Comment by joshmoore** [comment]
Yeah, I gave it a quick look. Definitely will require a little more Spring fu.

**Comment by chris-allan** [comment]
@joshmoore: I had tons of problems getting logging to behave like the previous version. Should now be all set with the additional commits. Also included now in this PR are an upgrade to the latest omero-blitz and some updates to the documentation.

**Comment by joshmoore** [comment]
re: logging -- :+1: 

The need to copy omero.properties is unfortunate, but is likely low enough churn at this point that it's not worth overthinking it.

**Pull request merged**